File: DS Week 3/Trial DS Week 3/BinarySearchTree.js
class Node{
    constructor(value){
        this.value=value
        this.rightChild=null
        this.leftChild=null
    }
}
class BinarySearchTree{
    constructor(){
        this.root=null
    }
    insertNode(value){
        let newNode=new Node(value)
        if(!this.root){
            this.root=newNode
            return
        }
        let currentNode=this.root
        while(currentNode){
            if(value<currentNode.value){
                if(!currentNode.leftChild){
                    currentNode.leftChild=newNode
                    break
                }
                currentNode=currentNode.leftChild
            }else{
                if(!currentNode.rightChild){
                    currentNode.rightChild=newNode
                    break
                }
                currentNode=currentNode.rightChild
            }
        }
    }
    insertNodeByRecursion(value){
        let newNode=new Node(value)
        if(!this.root){
            this.root=newNode
            return
        }
        this.insertNodeByRecursionHelper(value,newNode,this.root)
    }
    insertNodeByRecursionHelper(value,newNode,current){
        if(value<current.value){
            if(!current.leftChild){
                current.leftChild=newNode
                return
            }
            this.insertNodeByRecursion(value,newNode,current.leftChild)
        }else{
            if(!current.rightChild){
                current.rightChild=newNode
                return
            }
            this.insertNodeByRecursion(value,newNode,current.rightChild)
        }
    }
    removeNode(value){
        this.removeNodeHelper(value,this.root,null)
    }
    removeNodeHelper(value,current,parent){
        while(current){
            if(value<current.value){
                parent=current
                current=current.leftChild
            }else if(value>current.value){
                parent=current
                current=current.rightChild
            }else{
                if(!current.leftChild && !current.rightChild){
                    if(current.value<parent.value){
                        parent.leftChild=null
                    }else{
                        parent.rightChild=null
                    }
                }
                else if(!current.rightChild && current.leftChild){
                    if(current.value<parent.value){
                        parent.leftChild=current.leftChild
                    }else{
                        parent.rightChild=current.leftChild
                    }
                }
                else{
                    current.value=current.rightChild.value
                    this.removeNodeHelper(current.rightChild.value,current.rightChild,current)
                }
            }
        }
        console.log('node does not exitst');
    }
    removeNodeRecurssion(value){
        this.removeNodeRecurssionHelper(value,this.root,null)
    }
    removeNodeRecurssionHelper(value,current){
        if(!current){
            console.log('node does not exitst');
            return
        }

        if(value<current.value){
            parent=current
            removeNodeRecurssionHelper(value,current.leftChild)
        }else if(value>current.value){
            parent=current
            removeNodeRecurssionHelper(value,current.rightChild)
        }else{
            if(!current.leftChild && !current.rightChild){
                if(current.value<parent.value){
                    parent.leftChild=null
                }else{
                    parent.rightChild=null
                }
            }
            else if(!current.rightChild && current.leftChild){
                if(current.value<parent.value){
                    parent.leftChild=current.leftChild
                }else{
                    parent.rightChild=current.leftChild
                }
            }
            else{
                current.value=current.rightChild.value
                this.removeNodeRecurssionHelper(current.rightChild.value,current.rightChild)
            }
        }
    }
    inOrder(){
        this.inOrderHelper(this.root)
    }
    inOrderHelper(current){
        if(!current){
            return
        }
        this.inOrderHelper(current.leftChild)
        console.log(current.value);
        this.inOrderHelper(current.rightChild)
    }
    preOrder(){
        this.preOrderHelper(this.root)
    }
    preOrderHelper(current){
        if(!current){
            return
        }
        console.log(current.value);
        this.preOrderHelper(current.leftChild)
        this.preOrderHelper(current.rightChild)
    }
    postOrder(){
        this.postOrderHelper(this.root)
    }
    postOrderHelper(current){
        if(!current){
            return
        }
        this.postOrderHelper(current.leftChild)
        this.postOrderHelper(current.rightChild)
        console.log(current.value);
    }
    levelOrder(){
        let queue=[]
        queue.push(this.root)
        while(queue.length>=1){
            let current=queue.shift()
            console.log(current.value);
            if(current.leftChild){
                queue.push(current.leftChild)
            }
            if (current.rightChild) {
                queue.push(current.rightChild)
            }
        }
    }

}

let bst=new BinarySearchTree()
bst.insertNode(10)
bst.insertNode(8)
bst.insertNode(7)
bst.insertNode(9)
bst.insertNode(12)
bst.insertNode(11)
bst.insertNode(13)
// bst.inOrder()
// bst.preOrder()
bst.levelOrder()

File: DS Week 3/Binary search tree/BinarySearchTree.java
public class BinarySearchTree{
    public static void main(String[] args) {
        BinarySearchTree binaryTree=new BinarySearchTree();
        binaryTree.addNode(10);
        binaryTree.addNode(9);
        binaryTree.addNode(11);
        binaryTree.addNode(12);
        binaryTree.addNode(14);
        binaryTree.addNode(13);
        binaryTree.addNode(17);
        // binaryTree.closest(16);
        // binaryTree.remove(11);
        binaryTree.inOrder();
        // binaryTree.preOrder();
        // binaryTree.postOrder();
        // // int a=5;
        // System.out.println(binaryTree.contains(a)); 
        
    }

    class Node{
        int data;
        Node left,right;
        Node (int data){
            this.data=data;
        }
    }

    public Node root=null;

    public void addNode(int data){
        Node newNode=new Node(data);
        Node currentNode=root;
        if(currentNode==null){
            root=newNode;
            return;
        }else{
            while(true){
                if(data<currentNode.data){
                    if(currentNode.left==null){
                        currentNode.left=newNode;
                        break;
                    }
                    else{currentNode=currentNode.left;}
                }else if(data>=currentNode.data){
                    if(currentNode.right==null){
                        currentNode.right=newNode;
                        break;
                    }
                    else{
                        currentNode=currentNode.right;
                    }
                }
            }
        }
    }

    public boolean contains(int data){
        Node currentNode=root;
        while(currentNode.data!=data){
            if(data<currentNode.data){
                currentNode=currentNode.left;
                if(currentNode==null){
                    return false;
                }
            }
            else if(data>=currentNode.data){
                currentNode=currentNode.right;
                if(currentNode==null){
                    return false;
                }
            }
            
        }
        return true;
        
    }

    public void remove(int data){
        Node currentNode=root;
        Node parentNode=null;
        removeHelper(currentNode, parentNode, data);
    }

    public void removeHelper (Node currentNode,Node parentNode,int data){
        if(currentNode==null){
            System.out.println("Empty");
            return;
        }
        while(currentNode!=null){
            
            if(data<currentNode.data){
                parentNode=currentNode;
                currentNode=currentNode.left;
            }
            else if(data>currentNode.data){
                parentNode=currentNode;
                currentNode=currentNode. right;
            }
            else{

                if(currentNode.left==null&&currentNode.right==null){
                    if(parentNode!=null){
                        if(currentNode==parentNode.left){
                            parentNode.left=null;
                            return;
                        }
                        else{
                            parentNode.right=null;
                            return;
                        }
                    }else{
                        currentNode=null;
                        root=null;
                        return;
                    }
                }else if(currentNode.right==null){
                    currentNode.data=getMaximum(currentNode.left);
                    removeHelper(currentNode.left,currentNode,currentNode.data);
                    return;
                    
                }else{
                    currentNode.data=getMinimum(currentNode.right);
                    removeHelper(currentNode.right,currentNode,currentNode.data);
                    return;
                }
            }
        }
        if(currentNode==null){
            System.out.println("No element found");
            return;
        }
    }

    public int getMinimum(Node curretNode){
        if(curretNode.left==null){
            return curretNode.data;

        }else{
            return getMinimum(curretNode.left);
        }
    }

    public int getMaximum(Node curretNode){
        if(curretNode.right==null){
            return curretNode.data;

        }else{
            return getMaximum(curretNode.right);
        }
    }
    
    public void inOrder(){
        inOrderHelper(root);
        System.out.println();
    }

    public void inOrderHelper(Node node){
        if(node!=null){
            inOrderHelper(node.left);
            System.out.print(node.data+" ");
            inOrderHelper(node.right);
        }
    }

    public void preOrder(){
        preOrderHelper(root);
        System.out.println();
    }

    public void preOrderHelper(Node node){
        if(node!=null){
            System.out.print(node.data+" ");
            preOrderHelper(node.left);
            preOrderHelper(node.right);
        }
    }

    public void postOrder(){
        postOrderHelper(root);
        System.out.println();
    }

    public void postOrderHelper(Node node){
        if(node!=null){
            postOrderHelper(node.left);
            postOrderHelper(node.right);
            System.out.print(node.data+" ");
        }
    }

    public void closest(int target){
        Node current=root;
        if(current==null){
            System.out.println("empty");
            return;
        }
        int closest=current.data;
        
        while(current!=null){
            if(Math.abs(target-current.data)<Math.abs(target-closest)){
                closest=current.data;
            }
            if(target<current.data){
                current=current.left;
            }else if(target>current.data){
                current=current.right;
            }else{
                closest=current.data;
                break;
            }
        }

        System.out.println(closest);
        return;
    }

}